texk/dvipos/dvipos.1
.TH "DVIPOS" "1" "30 August 2022" "dvipos 20070107 (KPATHSEA)" "ConTeXt"
.de URL
\\$2 \(laURL: \\$1 \(ra\\$3
..
.if \n[.g] .mso www.tmac
.de EX
.in +3
.nf
.ft CW
..
.de EE
.in -3
.ft R
.fi
..

.SH NAME
dvipos \- compute positions in a DVI file

.SH SYNOPSIS
\fBdvipos\fP [ \fIOPTION\fP... ] \fIinfile[.dvi]\fP

.SH DESCRIPTION
\fBdvipos\fP parses a DVI file looking for \fIpos:\fP specials.  It
currently recognizes \fIpos:pxy\fP, \fIpos:pxywhd\fP,
\fIpos:pxyplus\fP, \fIpos:begbox\fP, \fIpos:endbox\fP,
\fIpos:beglines\fP, and \fIpos:endlines\fP.  It then outputs the
information from those specials along with information that only a DVI
postprocessor could determine, such as the current \fIx\fP and \fIy\fP
location.  The output looks like:
.EX
\\pospxywhd{text:54}{54}{10663sp}{8535sp}{21326sp}{34081sp}{0sp}
.EE
and is suitable for including in a 
.BR tex (1)
or
.BR pdftex (1)
auxiliary file, such as \fIdocument.tuo\fP, to be read in by the next
run of the
.BR tex (1)
or similar typesetting engine.

.SH OPTIONS

.TP
\fB-h, --help\fR
print usage.
.TP
\fB-W, --warning\fR
print warnings.
.TP
\fB-v, --verbose\fR
print verbose output.
.TP
\fB-d, --debug\fR
print 
.BR dvitype (1)
debugging data.
.TP
\fB-s, --smashchars\fR
regard height and depth as zero.
.TP
\fB-o, --output=\fIFILE\fR
send all output to FILE.  Without this option, output goes to stdout.
.TP
\fB-b, --bbox[=\fP\fIFILE\fP]\fR
send bounding box to FILE (default FILE is \fIinfile.pos\fP).
.TP
\fB-f, --framed[=\fP\fIBASE\fP]\fR
request copy of DVI file, \fIBASE.dvi\fP, with bounding boxes
\fBframed\fP (default FILE is \fIinfile_frames.dvi\fP).  See the
\fB--framesize\fP option.
.TP
\fB-m, --mag=\fIINT\fR
override 
.BR tex (1)
magnification by INT.
.TP
\fB-w, --framesize=\fIINT\fR
set frame rule size by INT (default 6554 = .1pt).
.TP
\fB-p, --pages=\fIX:Y\fR
set page ranges from X to Y.

.SH USAGE
.B dvipos
is run mostly behind the scenes by ConTeXt's
.BR texexec (1)
between runs of 
.BR tex (1)
or
.BR pdftex (1),
in order to provide information to and get information about
positional graphics.  These graphics are provided by the MetaFun
interface to
.BR mpost (1).
\fBdvipos\fP is not used (or needed) in PDF-output mode but is
needed for DVI-output mode.  
.BR pdftex (1)
starting with version 1.40 can produce position information even in
DVI mode, so once the ConTeXt macros take advantage of this feature,
the need for \fBdvipos\fP will fade.

Until then, here is a typical use of \fBdvipos\fP:
.EX
dvipos notes.dvi
.EE
The position information that is output is collected by
.BR texexec (1)
to include in the \fInotes.tuo\fP auxiliary file to use in the next
run.


.SH EXAMPLES
.TP
Extract positions from \fIdoc.dvi\fP, with output to \fIdoc.loc\fP:
\f(CWdvipos -o doc.loc doc.dvi\fP
.TP
Same as above, but also send bbox information to \fIdoc.pos\fP:
\f(CWdvipos -b -o doc.loc doc.dvi\fP
.TP
Instead send bbox information to \fI/tmp/debug.pos\fP:
\f(CWdvipos -b=/tmp/debug.pos -o doc.loc doc.dvi\fP

.SH "SEE ALSO"
.BR dvips (1),
.BR dvitype (1),
.BR mpost (1),
.BR pdftex (1),
.BR tex (1),
.BR texexec (1).

For more about positional graphics, see the MetaFun documentation on
the
.URL "https://wiki.contextgarden.net/MetaFun" "ConTeXt wiki" .

.SH BUGS
If you specify a file for the \fB-b\fP option, you must join the
option and name with \fB=\fP rather than (the more common) space:
.EX
dvipos -b=/tmp/debug.pos notes.dvi
.EE
works, but
.EX
dvipos -b /tmp/debug.pos notes.dvi
.EE
does not.

.SH AUTHOR
.B dvipos
is written by Jin-Hwan Cho <chofchof@ktug.or.kr>.  It is free
software (GPLv2 or later).  This manpage was written by Sanjoy
Mahajan <sanjoy@mit.edu> and is in the public domain.  It is currently
maintained as part of TeX Live.


ChangeLog
2022-03-21  Karl Berry  <karl@freefriends.org>

	* tardate.ac (tex_live_tardate): 2020-03-21 for TL'22.

2022-01-18  Karl Berry  <karl@freefriends.org>

	* version.ac: 2022, no more /dev.

2021-10-04  Karl Berry  <karl@freefriends.org>

	* reautoconf with automake-1.16.5.

2021-06-27  Karl Berry  <karl@freefriends.org>

	* reautoconf: pass -Wno-obsolete by default, since Autoconf 2.70
	and later obsoletes many common macros. See m4/README for more.
	Also standardize shell usage, etc., but no other functional changes.

2021-04-17  Karl Berry  <karl@tug.org>

	* version.ac: now 2022/dev since TL'21 is released.

2021-03-23  Karl Berry  <karl@tug.org>

	* tardate.ac (tex_live_tardate): 2020-03-23 for TL'21.

2021-02-17  Karl Berry  <karl@tug.org>

	* version.ac: 2021, no more /dev.

2020-05-28  Karl Berry  <karl@freefriends.org>

	* Makefile.am (world): do not descend into $(texlinks_dir) if it
	doesn't exist. Report from Andreas Scherer,
	https://tug.org/pipermail/tlbuild/2020q1/004614.html.

2020-05-14  Karl Berry  <karl@freefriends.org>

	* reautoconf for removal of poppler support.

2020-04-10  Karl Berry  <karl@tug.org>

	* version.ac: 2020/dev.

2020-03-27  Karl Berry  <karl@tug.org>

	* tardate.ac (tex_live_tardate): 2020-03-27 for TL'20.

2020-03-06  Karl Berry  <karl@tug.org>

	* .travis.yml: attempt to reenable musl build on github.

2020-01-31  Karl Berry  <karl@tug.org>

	* Makefile.am (texlinks_dir): now texk/texlive/tl_support
	(instead of tl_scripts).

2020-01-16  Karl Berry  <karl@freefriends.org>

	* version.ac (tex_live_version): remove /dev, preparing for TL'20.

2019-06-25  Karl Berry  <karl@freefriends.org>

	* Build: don't call buildinfo from here, the output when the CC
	etc. envvars are not set is too misleading. Just write UNAME.

2019-06-24  Karl Berry  <karl@freefriends.org>

	* doc/buildinfo.sh (compiler_version): new shell function;
	 ignore useless first "Configured:" line for gcc->cc on Macs.

2019-06-23  Karl Berry  <karl@freefriends.org>

	* doc/buildinfo.sh: output the compiler/other names
	that we pass from the buildinfo target.
	* Makefile.am (buildinfo): pass OBJCXX, not OBJC.
	* Build: more maintainable exports of TL_*.

2019-06-22  Karl Berry  <karl@freefriends.org>

	* doc/buildinfo.sh: rename from build-info.sh;
	forget /etc/issue;
	allow option to omit config.status output;
	do not dump environment;
	run lsb_release -a. We'll see ...
	* Build: use buildinfo.sh instead of inline commands (mostly).
	* Makefile.am (buildinfo): make dependency of all-recursive
	so that it runs first; rename from build-info; remove from world.

2019-06-06  Karl Berry  <karl@tug.org>

	* Makefile.am (world): make build-info.
	(build-info): new target to invoke doc/build-info.sh.
	(build_info_target): new variable.
	* Build: export TL_* variables so build-info.sh can report them.
	* doc/build-info.sh: new script to report compiler versions,
	dump environment, config.status, etc.

2019-05-03  Karl Berry  <karl@tug.org>

	* version.ac: 2020/dev.

2019-04-07  Karl Berry  <karl@freefriends.org>

	* tardate.ac (tex_live_tardate): let's hope for 2019-04-07.

2018-10-02  Karl Berry  <karl@tug.org>

	* Build: TL_MAKE_VERBOSE=VERBOSE=1 by default,
	so that test-suite.log will be shown on failure.
	https://lists.gnu.org/archive/html/automake/2018-09/msg00015.html

2018-04-28  Karl Berry  <karl@tug.org>

	* version.ac: 2019/dev.

2018-04-10  Karl Berry  <karl@tug.org>

	* README.*: update from new tlbuild.

2017-11-15  Karl Berry  <karl@tug.org>

	* *.test: use #!/bin/sh -vx. Add 

2017-03-21  Karl Berry  <karl@freefriends.org>

	* everywhere: use EXTRA_DIST=TLpatches instead of <pkgname>-PATCHES.

2017-03-18  Karl Berry  <karl@tug.org>

	* version.ac (tex_live_version): 2017/dev => 2017 for TL17.

2016-11-16  Karl Berry  <karl@tug.org>

	* Makefile.am (install-exec-hook): warn that the
	result of make install is not in itself usable.

2016-06-30  Karl Berry  <karl@tug.org>

	* version.ac (tex_live_version): 2016 => 2017/dev.
	* tardate.ac (tex_live_tardate): 2015-05-12 => 2016-05-23.
	(Though this is not used for any release.)

2016-03-16  Karl Berry  <karl@tug.org>

	* Build: check for exitstatus.txt not having an exit status.

2016-03-06  Karl Berry  <karl@tug.org>

	* version.ac: remove /dev to prepare for 2016.

2015-06-12  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* version.ac: Switch 2015 => 2016/dev.
	* tardate.ac: Switch to 2015-06-12.

2015-05-28  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* README.solaris (removed): Moved from here to doc/.
	* Makefile.am: Adapted.

2015-04-14  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* README.solaris: New file from Mojca (without very long lines).
	* Makefile.am: Add the new file.

2015-04-12  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* version.ac (tex_live_version): starting pretest, no more /dev.

2015-03-07  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* tardate.ac: Switch to 2015-03-07.

2014-11-24  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* tardate.ac: Switch to 2014-11-24.

2014-10-21  Karl Berry  <karl@tug.org>

	* configure.ac: report CC/CXX values, with --version output
	if available.
	* Build: no need for objc any more.

2014-08-22  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* tardate.ac: Switch to 2014-08-22.

2014-06-16  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* version.ac: Switch 2014 => 2015/dev.
	* tardate.ac: Switch to 2014-06-16.

2014-04-28  Peter Breitenlohner  <peb@mppmu.mpg.de>

	Makefile.am: New names for README.* files.

2014-04-17  Karl Berry  <karl@tug.org>

	* version.ac (tex_live_version): starting pretest, no more /dev.

2014-04-04  Karl Berry  <karl@tug.org>

	* Build (TL_MAKE,MAKE): set both in all cases.
	(buildenv.log): include pwd, and use .log instead of .txt.

2014-04-01  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* tardate.ac: Switch to 2014-04-01.

2014-01-01  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* tardate.ac: Switch to 2014-01-01.

2013-12-26  Peter Breitenlohner  <peb@mppmu.mpg.de>


texk/dvipos/dvipos.c
/*  dvipos-20070107

    Copyright (C) 2003 by Jin-Hwan <chofchof@ktug.or.kr>
    
    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.
    
    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.
    
    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA.
*/

#ifdef HAVE_CONFIG_H
#include <config.h>
#else
#define PACKAGE_BUGREPORT "tex-k@tug.org"
#endif

#include "utils.h"
#include "dvicore.h"

#include <limits.h>
#ifdef MIKTEX
#include <getopt.h>
#endif

/* Global variables */
int verbose = M_FAIL, do_smashchars = 0; /* output message only if failed */
FILE *infp = NULL, *outfp = NULL, *bbxfp = NULL, *frmfp = NULL;
char *infname = NULL, *bbxfname = NULL, *frmfname = NULL;
SIGNED_QUAD dbg_location = 0, rule_width = 6554; /* .1pt */

/* Internal variables */
static char *outfname = NULL;
static char deffname[] = "default";
static int nmag = 0, resol = 300, bpage = 1, epage = INT_MAX;

#ifdef MIKTEX
#define PLATFORM "(MiKTeX 2.2)"
#else
#define PLATFORM "(KPATHSEA)"
#endif

static void title (void)
{
  msg_out(M_VERBOSE, "This is DVIpos, Version %s %s\nby Jin-Hwan Cho <chofchof@ktug.or.kr>\n\n", VERSION, PLATFORM);
}

static void usage (void)
{
  fprintf(stdout, "Usage: dvipos [OPTION...] infile[.dvi]\n\n"
                  "-h,  --help           print usage\n"
                  "-W,  --warning        print warnings\n"
                  "-v,  --verbose        print verbose output\n"
                  "-d,  --debug          print 'dvitype' debugging data\n"
                  "-s,  --smashchars     regard height and depth as zero\n"
                  "-o,  --output=FILE    send all output to FILE\n"
                  "-b,  --bbox[=FILE]    send bounding box to FILE (default FILE is infile.pos)\n"
                  "-f,  --framed[=BASE]  request copy of DVI file, BASE.dvi, with bounding boxes\n"
		  "                      *framed* (default FILE is infile_frames.dvi)\n"
                  "-m,  --mag=INT        override TeX magnification by INT\n"
//                "-r,  --resolution=INT set resolution by INT\n"
                  "-w,  --framesize=INT   set frame rule size by INT (default 6554 = .1pt)\n"
                  "-p,  --pages=X:Y      set page ranges from X to Y\n"
                  "\nEmail bug reports to %s.\n", PACKAGE_BUGREPORT);
}

static struct option long_options[] = {
  /* Options without arguments: */
  { "debug", no_argument, NULL, 'd' },
  { "help", no_argument, NULL, 'h' },
  { "verbose", no_argument, NULL, 'v' },
  { "warning", no_argument, NULL, 'W' },
  { "smashchars", no_argument, NULL, 's' },
  /* Options accepting an argument: */
  { "mag", required_argument, NULL, 'm' },
  { "framesize", required_argument, NULL, 'w' },
  { "pages", required_argument, NULL, 'p' },
  { "resolution", required_argument, NULL, 'r' },
  { "output", optional_argument, NULL, 'o' },
  /* Options accepting an optional argument: */
  { "bbox", optional_argument, NULL, 'b' },
  { "framed", optional_argument, NULL, 'f' },
  { 0, 0, 0, 0 }
};

static void read_options (int argc, char **argv)
{
  int op;

  while ((op = getopt_long(argc, argv, "dhvwm:s:r:p:o:b::f::", long_options, (int *)0)) != EOF) {
    char *p, *q;
    switch (op) {
    case 'd': /* debug */
      verbose |= M_DEBUG;
      break;
    case 'h': /* help */
      verbose |= M_VERBOSE; title(); usage(); exit(1);
      break;
    case 'v': /* verbose */
      verbose |= M_VERBOSE;
      break;
    case 'W': /* warning */
      verbose |= M_WARNING;
      break;
    case 's': /* smashchars */
      do_smashchars = 1;
      break;
    case 'm': /* mag */
      if (*(q = optarg) == '=') q++;
      nmag = (int)strtol(q, &p, 0);
      if (p == q || nmag < 0)
        msg_out(M_FAIL, "Missing a number for magnification.\n");
      break;
    case 'r': /* resolution */
      if (*(q = optarg) == '=') q++;
      resol = (int)strtol(q, &p, 0);
      if (p == q || resol < 0)
        msg_out(M_FAIL, "Missing a number for resolution.\n");
      break;
    case 'w': /* framesize */
      if (*(q = optarg) == '=') q++;
      rule_width = (int)strtol(q, &p, 0);
      if (p == q || rule_width < 0)
        msg_out(M_FAIL, "Missing a number for frame rule size.\n");
      break;
    case 'p': /* pages */
      if (*(q = optarg) == '=') q++;
      bpage = (int)strtol(q, &p, 0);
      if (p == q) msg_out(M_FAIL, "Missing a page range.\n");
      if (*p == ':') {
        epage = (int)strtol(q = p + 1, &p, 0);
	if (p == q) msg_out(M_FAIL, "Missing a 2nd page number.\n");
      }
      if (bpage > epage) {
        int sw = bpage; bpage = epage; epage = sw;
      }
      break;
    case 'o': /* output */
      if (*(q = optarg) == '=') q++;
      if (strlen(q) > 0) outfname = xstrdup(q);
      else msg_out(M_FAIL, "Missing an output file name.\n");
      break;
    case 'b': /* bbox */
      bbxfname = deffname;
      if (optarg) {
        if (*(q = optarg) == '=') q++;
        if (strlen(q) > 0) bbxfname = xstrdup(q);
      }
      break;
    case 'f': /* framed */
      frmfname = deffname;
      if (optarg) {
        if (*(q = optarg) == '=') q++;
        if (strlen(q) > 0) frmfname = xstrdup(q);
      }
      break;
    default:
      msg_out(M_FAIL, "Unparsed option (%d)\n", op);
    }
  }

  /* The first non-optional argument is to be used as the DVI file name */
  if (optind < argc) {
    infname = make_suffix(argv[optind], "dvi");
    optind++;
    while (optind < argc)
      msg_out(M_WARNING, "Warning: Ignore unparsed argument (%s).\n", argv[optind++]);
  }

  /* set up all the options */
  msg_out(M_DEBUG, "Options selected:\n");
  msg_out(M_DEBUG, "  Starting page = * \n");
  msg_out(M_DEBUG, "  Maximum number of pages = 1000000\n");
  msg_out(M_DEBUG, "  Output level = 4 (the works)\n");
  msg_out(M_DEBUG, "  Resolution = %d.00000000 pixels per inch\n", resol);
  if (nmag > 0)
    msg_out(M_DEBUG, "  New magnification factor = %-8.03f\n", nmag / 1000.);
}

int main (int argc, char *argv[]) 
{
  int i;

  if (argc > 1)
    read_options(argc, argv);

  title();

  if (infname == NULL)
    msg_out(M_FAIL, "No DVI filename specified.\n");

//  if (outfname == NULL)
//    outfname = make_suffix(infname, "tuo");

  if (bbxfname == deffname)
    bbxfname = make_suffix(infname, "pos");

  if (frmfname) {
    if (frmfname == deffname) {
      frmfname = make_suffix(infname, "frames.dvi");
      *(frmfname+strlen(frmfname)-11) = '_';
    } else {
      char *p = frmfname;
      frmfname = make_suffix(p, "dvi");
      free(p);
    }
  }

  /* Open the input DVI file */
  if ((infp = fopen(infname, "rb")) == NULL)
    msg_out(M_FAIL, "Could not open the input file %s.", infname);

  /* Open the output file */
  if (outfname == NULL) {
    outfp = stdout;
#ifdef WIN32
    setmode(fileno(stdout), _O_BINARY);
#endif
  } else if ((outfp = fopen(outfname, "ab")) == NULL)
    msg_out(M_FAIL, "Could not open the output file %s.\n", outfname);

  /* Open the boundingbox file */
  if (bbxfname && (bbxfp = fopen(bbxfname, "wb")) == NULL)
    msg_out(M_FAIL, "Could not open the boundingbox file %s.\n", bbxfname);

  /* Open the framed DVI file */
  if (frmfname && (frmfp = fopen(frmfname, "wb")) == NULL)
    msg_out(M_FAIL, "Could not open the framed DVI file %s.\n", frmfname);

#ifdef KPATHSEA
  kpse_set_program_name(argv[0], "dvipos");
#elif MIKTEX
  miktex_initialize();
#endif

  dvi_init(nmag, resol);

  if (bpage <= 0) bpage = 1;
  if (epage >= dvi_pages) epage = dvi_pages;

  for (i = bpage; i <= epage; i++)
    dvi_do_page(i);

  dvi_close();

  if (infname) free(infname);
  if (outfname) free(outfname);
  if (bbxfname) free(frmfname);
  if (frmfname) free(frmfname);

  if (outfp) fclose(outfp);
  if (bbxfp) fclose(bbxfp);
  if (frmfp) fclose(frmfp);

#ifdef MIKTEX
  miktex_uninitialize();
#endif

  exit(0);
}
